### PR TITLE
Selectable Api Server endpoints aka ExposureClasses [Part 1+2]

### PR DESCRIPTION
**Pull request by dkistner**
**How to categorize this PR?**
/area control-plane open-source
/kind api-change enhancement

**What this PR does / why we need it**:
This PR implements the **first** and **second** step of the [Selectable Api Server endpoint story](https://github.com/gardener/gardener/issues/3505). This story will be implemented in several steps, see: https://github.com/gardener/gardener/issues/3505#issuecomment-821118819

This PR introduces a new resource called `ExposureClass` to the `core` api group (currently only in `v1alpha1`).
Please check the [implementation proposal](https://github.com/gardener/gardener/issues/3505#issuecomment-795656500) for the details.

In detail this PR adds the following:
- For step 1:
  - The api definition for the `ExposureClass` resource
  - The ExposureClass reference in the Shoot resource
  - The Api Validation for the ExposureClass resource
  - An resource reference check for ExposureClasses referenced in Shoot resource via the `resourcereference` admission plugin
  - The ExposureClass storage implementation
  - An ExposureClass controller implementation for the `gardener-controller-manager` (adding finalizers and removing them when no Shoot reference the ExposureClass anymore)
- For step 2:
  - The `ExposureClassHandler` api in the GardenletConfiguration + api validation
  - The Gardenlet's shoot controller applies the handler config for a Shoot with matching `ExposureClass` if no SNI is enabled

**Note:** Referencing an `ExposureClass` in Shoot resources  via `.spec.exposureClassName` is temporarily disabled via validation. This will be enabled again when the implementation of the story has been completed.

**Which issue(s) this PR fixes**:
Step 1+2 of https://github.com/gardener/gardener/issues/3505

**Special notes for your reviewer**:
Unfortunately the PR is quite large. There are a lot of generations for the ExposureClass, Shoot and GardenletConfiguration apis, clients, mocks etc. The subsequent PR for the ExposureClass story will be most likely smaller.

**Release note**:

```NONE

```

### CODE REVIEW

File: pkg/apis/core/types_exposureclass.go [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ExposureClass represents a control plane endpoint exposure strategy.
+type ExposureClass struct {
+	metav1.TypeMeta
+	// Standard object metadata.
+	metav1.ObjectMeta
+	// Handler is the name of the handler which appies the control plane endpoint exposure strategy.
```
Comment (rfranzke) [inline pkg/apis/core/types_exposureclass.go]:
```suggestion
	// Handler is the name of the handler which applies the control plane endpoint exposure strategy.
```

(same for other occurrences)

File: pkg/apis/core/types_exposureclass.go [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+	Handler string
+	// Scheduling holds information to select applicable Seed's for ExposureClass usage.
```
Comment (rfranzke) [inline pkg/apis/core/types_exposureclass.go]:
```suggestion
	// Scheduling holds information how to select applicable Seed's for ExposureClass usage.
```

(same for other occurrences)

File: pkg/apis/core/v1beta1/types_exposureclass.go [hunk 3]
```diff
@@ -0,0 +1,60 @@
+// Copyright (c) 2021 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
```
Comment (rfranzke) [inline pkg/apis/core/v1beta1/types_exposureclass.go]: Similar to https://github.com/gardener/gardener/pull/3878, we probably should only add it to the `v1alpha1` version for now and promote it to `v1beta1` later if we are confident enough with the API?

File: pkg/apis/core/v1beta1/types_shoot.go [hunk 4]
```diff
@@ -115,6 +115,9 @@ type ShootSpec struct {
 	// +patchStrategy=merge
 	// +optional
 	Tolerations []Toleration `json:"tolerations,omitempty" patchStrategy:"merge" patchMergeKey:"key" protobuf:"bytes,17,rep,name=tolerations"`
+	// ExposureClassName is the optional name of an exposure class to apply a control plane endpoint exposure strategy.
+	// +optional
+	ExposureClassName *string `json:"exposureClassName,omitempty" protobuf:"bytes,18,opt,name=exposureClassName"`
```
Comment (rfranzke) [inline pkg/apis/core/v1beta1/types_shoot.go:120]: Can you please enhance https://github.com/gardener/gardener/blob/master/example/90-shoot.yaml with this new field?

File: pkg/controllermanager/apis/config/v1alpha1/types.go [hunk 5]
```diff
@@ -61,6 +61,8 @@ type ControllerManagerControllerConfiguration struct {
 	// Event defines the configuration of the Event controller.  If unset, the event controller will be disabled.
 	// +optional
 	Event *EventControllerConfiguration `json:"event,omitempty"`
+	// ExposureClass defines the configuration of the ExposureClass controller.
+	ExposureClass *ExposureClassControllerConfiguration `json:"exposureClass,omitempty"`
```
Comment (rfranzke) [inline pkg/controllermanager/apis/config/v1alpha1/types.go:71]: Please enhance https://github.com/gardener/gardener/blob/master/example/20-componentconfig-gardener-controller-manager.yaml and https://github.com/gardener/gardener/blob/master/charts/gardener/controlplane/charts/runtime/templates/controller-manager/configmap-componentconfig.yaml

File: pkg/apis/core/validation/exposureclass.go [hunk 6]
```diff
@@ -17,17 +17,40 @@ package validation
...
+
+	return append(allErrs, validateExposureClassScheduling(exposureClass.Scheduling, field.NewPath("scheduling"))...)
+}
+
+func validateExposureClassScheduling(scheduling *core.ExposureClassScheduling, fldPath *field.Path) field.ErrorList {
+	var allErrs = field.ErrorList{}
+
+	if scheduling == nil {
+		return append(allErrs, field.Required(fldPath, "exposure class need to have a scheduling configuration"))
+	}
```
Comment (rfranzke) [inline pkg/apis/core/validation/exposureclass.go]: Why?
If it's not optional then the `ExposureClassScheduling` should not be a pointer in the API

File: pkg/apis/core/validation/exposurceclass_test.go [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+package validation_test
+
+import (
+	"github.com/gardener/gardener/pkg/apis/core"
+	. "github.com/gardener/gardener/pkg/apis/core/validation"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	. "github.com/onsi/gomega/gstruct"
+
```
Comment (rfranzke) [inline pkg/apis/core/validation/exposurceclass_test.go]:
```suggestion
```

File: pkg/apis/core/validation/exposureclass.go [hunk 8]
```diff
@@ -17,17 +17,40 @@ package validation
...
+
+	if (scheduling.SeedSelector == nil && len(scheduling.Tolerations) == 0) || (scheduling.SeedSelector != nil && len(scheduling.Tolerations) > 0) {
```
Comment (rfranzke) [inline pkg/apis/core/validation/exposureclass.go]: What about properly validating the tolerations like for `Shoot`s (https://github.com/gardener/gardener/blob/f5de6826ec488d56f37566a71613b0ae90a414b7/pkg/apis/core/validation/shoot.go#L156)?

Comment (rfranzke) [inline pkg/apis/core/validation/exposureclass.go]: Why is this constraint, i.e., why don't we allow having `ExposureClass`es with neither a seed selector/tolerations?

File: pkg/gardenlet/apis/config/v1alpha1/types.go [hunk 9]
```diff
@@ -83,6 +83,9 @@ type GardenletConfiguration struct {
 	// by the Gardenlet in the seed clusters.
 	// +optional
 	SNI *SNI `json:"sni,omitempty"`
+	// ExposureClassHandlers is a list of optional of exposure class handlers.
+	// +optional
+	ExposureClassHandlers []ExposureClassHandler `json:"exposureClassHandlers,omitempty"`
```
Comment (rfranzke) [inline pkg/gardenlet/apis/config/v1alpha1/types.go:88]: Please enhance https://github.com/gardener/gardener/blob/master/example/20-componentconfig-gardenlet.yaml and https://github.com/gardener/gardener/blob/master/charts/gardener/gardenlet/charts/runtime/templates/configmap-componentconfig.yaml

File: pkg/gardenlet/apis/config/v1alpha1/types.go [hunk 10]
```diff
@@ -458,6 +461,27 @@ type SNIIngress struct {
...
+	// +optional
+	SNI *SNI `json:"sni,omitempty"`
+}
+
+// LoadBalancerService contains configuration which is used to configure the underlying
+// load balancer to apply the control plane endpoint exposure strategy.
+type LoadBalancerServiceConfig struct {
+	// Annotations is a key value map to annotate the underlying load balancer services.
+	// +optional
+	Annotations map[string]string `json:"annotations,omitempty"`
```
Comment (rfranzke) [inline pkg/gardenlet/apis/config/v1alpha1/types.go]: Why is it optional? Should the whole `LoadBalancerService` be optional then? 👀

File: pkg/gardenlet/apis/config/validation/validation.go [hunk 11]
```diff
@@ -74,6 +74,17 @@ func ValidateGardenletConfiguration(cfg *config.GardenletConfiguration, fldPath
 		}
 	}
 
+	exposureClassHandlersPath := fldPath.Child("exposureClassHandlers")
+	for i, handler := range cfg.ExposureClassHandlers {
+		handlerPath := exposureClassHandlersPath.Index(i)
+		if len(handler.Name) == 0 {
+			allErrs = append(allErrs, field.Invalid(handlerPath.Child("name"), handler.Name, "handler name cannot be empty"))
+		}
+		if len(handler.LoadBalancerService.Annotations) == 0 {
```
Comment (rfranzke) [inline pkg/gardenlet/apis/config/validation/validation.go]: OK, now I'm confused :) If it cannot be empty then above questions don't make sense anymore. You should rather just remove the 'optional' tag of this `Annotations` field.

**Review by rfranzke — CHANGES_REQUESTED** [review]
Very nice and well-structured PR, thanks! :)

File: pkg/apis/core/v1beta1/types_exposureclass.go (hunk 3 above)
Comment (dkistner) [reply pkg/apis/core/v1beta1/types_exposureclass.go]: Agree. I will remove the `v1beta1` api for now.

File: pkg/apis/core/v1beta1/types_shoot.go (hunk 4 above)
Comment (dkistner) [reply pkg/apis/core/v1beta1/types_shoot.go:120]: Yes I can, but I planned to provide docs and examples in a subsequent PR as last step of the story.

Currently it is forbidden via validation that a Shoot can reference an ExposureClass. The reason for this is that I wanted to split the story into several steps/PRs (too large at once, different reviewers are required for the different steps etc.). With the last step I will remove this validation again and finalize the feature by adding docs/examples etc. 

The reason why I did not add it in the examples is that the feature is not yet usable and I therefore don't wanted to advertise it. But if you want I will of course add the examples already now :)

File: pkg/apis/core/validation/exposureclass.go (hunk 8 above)
Comment (dkistner) [reply pkg/apis/core/validation/exposureclass.go]: I agree we should allow ExposureClasses without `scheduling`. Those ExposureClasses would select all available Seeds. Then I guess the `scheduling` section should stay a pointer and be optional.

File: pkg/apis/core/v1beta1/types_shoot.go (hunk 4 above)
Comment (rfranzke) [reply pkg/apis/core/v1beta1/types_shoot.go:120]: Ah, OK, I see, thanks. Sure, if you prefer to only enhance them later then when the feature is ready then it's fine for me. Let's just not forget it :)

File: pkg/gardenlet/apis/config/v1alpha1/types.go (hunk 10 above)
Comment (dkistner) [reply pkg/gardenlet/apis/config/v1alpha1/types.go]: No, it shouldn't be optional. Will remove it plus the belonging validation.

File: pkg/apis/core/types_exposureclass.go (hunk 1 above)
Comment (dkistner) [reply pkg/apis/core/types_exposureclass.go]: done

File: pkg/apis/core/types_exposureclass.go (hunk 2 above)
Comment (dkistner) [reply pkg/apis/core/types_exposureclass.go]: done

File: pkg/apis/core/v1beta1/types_exposureclass.go (hunk 3 above)
Comment (dkistner) [reply pkg/apis/core/v1beta1/types_exposureclass.go]: `v1beta1` is removed.

File: pkg/controllermanager/apis/config/v1alpha1/types.go (hunk 5 above)
Comment (dkistner) [reply pkg/controllermanager/apis/config/v1alpha1/types.go:71]: The chart is updated. As said I will add examples in a subsequent pr.

File: pkg/apis/core/validation/exposureclass.go (hunk 6 above)
Comment (dkistner) [reply pkg/apis/core/validation/exposureclass.go]: The `scheduling` section is now optional. Removed the validation.

File: pkg/apis/core/validation/exposureclass.go (hunk 8 above)
Comment (dkistner) [reply pkg/apis/core/validation/exposureclass.go]: done

Comment (dkistner) [reply pkg/apis/core/validation/exposureclass.go]: Removed this validation as `scheduling` is now optional.

File: pkg/gardenlet/apis/config/v1alpha1/types.go (hunk 10 above)
Comment (dkistner) [reply pkg/gardenlet/apis/config/v1alpha1/types.go]: done

File: pkg/gardenlet/apis/config/validation/validation.go (hunk 11 above)
Comment (dkistner) [reply pkg/gardenlet/apis/config/validation/validation.go]: see above. done

**Comment by dkistner** [comment]
@rfranzke Feedback is incorporated.

The `v1beta1` api version of `ExposureClass` is removed.
In the `resourcereferencemanager` admission plugin we need an informer/lister for ExposureClasses to check if a referenced ExposureClass of a Shoot exists. We are using the internalversion lister here and this worked well before with the `v1beta1` version of the ExposureClass. Now with the removal of the `v1beta1` I'm getting these kind of errors from the plugin:

```
E0507 15:52:37.830286   45221 reflector.go:138] github.com/gardener/gardener/pkg/client/core/informers/internalversion/factory.go:117: Failed to watch *core.ExposureClass: failed to list *core.ExposureClass: the server could not find the requested resource
```
Any idea why the internalversion informer/lister does not work with the `v1alpha1` api of ExposureClass?

Alternatively we could use the versioned informer/lister, see here: https://github.com/dkistner/gardener/commit/2c0f68338ad628e8dd681a312e730fb19b653746

File: pkg/gardenlet/apis/config/v1alpha1/types.go (hunk 9 above)
Comment (dkistner) [reply pkg/gardenlet/apis/config/v1alpha1/types.go:88]: I updated the chart. As said the examples will follow in a subsequent pr.

File: pkg/controllermanager/apis/config/v1alpha1/types.go [hunk 12]
```diff
@@ -64,6 +64,8 @@ type ControllerManagerControllerConfiguration struct {
 	// Event defines the configuration of the Event controller.  If unset, the event controller will be disabled.
 	// +optional
 	Event *EventControllerConfiguration `json:"event,omitempty"`
+	// ExposureClass defines the configuration of the ExposureClass controller.
```
Comment (rfranzke) [inline pkg/controllermanager/apis/config/v1alpha1/types.go:69]:
```suggestion
	// ExposureClass defines the configuration of the ExposureClass controller.
	// +optional
```

File: pkg/gardenlet/controller/shoot/shoot_control.go [hunk 13]
```diff
@@ -277,6 +277,7 @@ func (c *Controller) initializeOperation(ctx context.Context, logger *logrus.Ent
 		WithCloudProfileObject(cloudProfile).
 		WithShootSecretFromReader(gardenClient.APIReader()).
 		WithProjectName(project.Name).
+		WithExposureClassFromReader(gardenClient.APIReader()).
```
Comment (rfranzke) [inline pkg/gardenlet/controller/shoot/shoot_control.go:284]: You need to extend the seed authorizer, similar to what @danielfoehrKn does as part of #3893 (changes in `pkg/admissioncontroller`). As the gardenlet only reads the `ExposureClass`es, your implementation will be easier and you don't need to enhance the `seedrestriction` admission plugin.

File: pkg/operation/shoot/shoot.go [hunk 14]
```diff
@@ -148,6 +152,22 @@ func (b *Builder) WithShootSecretFromReader(c client.Reader) *Builder {
...
+// the exposure class with the given API reader.
+func (b *Builder) WithExposureClassFromReader(c client.Reader) *Builder {
+	b.exposureClassFunc = func(ctx context.Context, exposureClassName string) (*gardencorev1alpha1.ExposureClass, error) {
+		exposureClass := &gardencorev1alpha1.ExposureClass{}
+		if err := c.Get(ctx, kutil.Key(exposureClassName), exposureClass); err != nil {
+			return nil, err
+		}
+		if exposureClass == nil {
+			return nil, fmt.Errorf("could not fetch exposure class %q", exposureClassName)
+		}
```
Comment (rfranzke) [inline pkg/operation/shoot/shoot.go]: I don't think this is needed, is it? In which case would `exposureClass` be `nil` while `err` is also `nil`?

File: pkg/registry/core/exposureclass/storage/storage.go [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+	}
+	return &REST{store}
+}
+
+// Implement ShortNamesProvider.
+var _ rest.ShortNamesProvider = &REST{}
+
+// ShortNames implements the ShortNamesProvider interface. Returns a list of short names for a resource.
+func (r *REST) ShortNames() []string {
+	return []string{"expclass"}
```
Comment (rfranzke) [inline pkg/registry/core/exposureclass/storage/storage.go]: What about

```suggestion
	return []string{"expcls"}
```

? 😄

File: plugin/pkg/global/resourcereferencemanager/admission.go [hunk 16]
```diff
@@ -138,7 +139,10 @@ func (r *ReferenceManager) SetInternalCoreInformerFactory(f coreinformers.Shared
 	controllerDeploymentInformer := f.Core().InternalVersion().ControllerDeployments()
 	r.controllerDeploymentLister = controllerDeploymentInformer.Lister()
 
-	readyFuncs = append(readyFuncs, seedInformer.Informer().HasSynced, shootInformer.Informer().HasSynced, cloudProfileInformer.Informer().HasSynced, secretBindingInformer.Informer().HasSynced, quotaInformer.Informer().HasSynced, projectInformer.Informer().HasSynced, controllerDeploymentInformer.Informer().HasSynced)
+	exposurClassInformer := f.Core().InternalVersion().ExposureClasses()
```
Comment (rfranzke) [inline plugin/pkg/global/resourcereferencemanager/admission.go]:
```suggestion
	exposureClassInformer := f.Core().InternalVersion().ExposureClasses()
```

Comment (rfranzke) [inline plugin/pkg/global/resourcereferencemanager/admission.go]: As discussed, let's use the `v1beta1` lister here.

**Review by rfranzke — CHANGES_REQUESTED** [review]

File: pkg/registry/core/exposureclass/storage/storage.go (hunk 15 above)
Comment (dkistner) [reply pkg/registry/core/exposureclass/storage/storage.go]: Both also ok? I like `expclass` more. Easier to remember. ;)

Comment (rfranzke) [reply pkg/registry/core/exposureclass/storage/storage.go]: Sure :)

File: pkg/controllermanager/apis/config/v1alpha1/types.go (hunk 12 above)
Comment (dkistner) [reply pkg/controllermanager/apis/config/v1alpha1/types.go:69]: done

File: pkg/gardenlet/controller/shoot/shoot_control.go (hunk 13 above)
Comment (dkistner) [reply pkg/gardenlet/controller/shoot/shoot_control.go:284]: I added handling for ExposureClasses to the `SeedAuthorizer`, only get request will be allowed.

File: pkg/operation/shoot/shoot.go (hunk 14 above)
Comment (dkistner) [reply pkg/operation/shoot/shoot.go]: Yep, right. The `err` will be not empty if the `exposureClass` does not exists. Removed it.

File: pkg/registry/core/exposureclass/storage/storage.go (hunk 15 above)
Comment (dkistner) [reply pkg/registry/core/exposureclass/storage/storage.go]: As said. I added both.

File: plugin/pkg/global/resourcereferencemanager/admission.go (hunk 16 above)
Comment (dkistner) [reply plugin/pkg/global/resourcereferencemanager/admission.go]: done

Comment (dkistner) [reply plugin/pkg/global/resourcereferencemanager/admission.go]: You meant the `v1alpha1` lister here? The `v1beta1` version does not have the `exposureClass` anymore. Done.

**Comment by dkistner** [comment]
@rfranzke I have have applied the feedback. Could you please have another look?
/ready-for-review
/touch

**Review by rfranzke — APPROVED** [review]
Very nice PR, thanks a lot for addressing the feedback!

/lgtm

**Pull request merged**